docs-internal/db.md

```diff
@@ -0,0 +1,364 @@
+        .get("SELECT type FROM my_table WHERE name = ?", "Pikachu")
+        .thenAccept(maybeRow -> maybeRow
+                .ifPresent(row -> exec
+                        .insert("INSERT INTO my_table VALUES(?, ?)", "Raichu", row.column("type"))
+                )
+        )
+).exceptionally(throwable -> <fail somehow>)
+```
+
+### Interceptor implementation
+
+Example interceptor (for tracing):
+```java
+public class DbTracing implements DbInterceptor {
+    @Override
+    public void statement(DbInterceptorContext interceptorContext) {
+        Context context = interceptorContext.context();
+        Tracer tracer = context.get(Tracer.class).orElseGet(GlobalTracer::get);
+
+        // now if span context is missing, we build a span without a parent
+        Tracer.SpanBuilder spanBuilder = tracer.buildSpan(interceptorContext.dbType() + ":" + interceptorContext.statementName());
+
+        context.get(SpanContext.class)
+                .ifPresent(spanBuilder::asChildOf);
+
+        Span span = spanBuilder.start();
+
+        interceptorContext.statementFuture().thenAccept(nothing -> span.log(CollectionsHelper.mapOf("type", "statement")));
+
+        interceptorContext.resultFuture().thenAccept(count -> span.log(CollectionsHelper.mapOf("type", "result",
+                                                                "count", count)).finish())
+                .exceptionally(throwable -> {
+                    Tags.ERROR.set(span, Boolean.TRUE);
+                    span.log(CollectionsHelper.mapOf("event", "error",
+                                                     "error.kind", "Exception",
+                                                     "error.object", throwable,
+                                                     "message", throwable.getMessage()));
+                    span.finish();
+                    return null;
+                });
+
+    }
+
+    public static DbTracing create() {
+        return new DbTracing();
+    }
+}
+```
+
+### Mapper implementation
+
+Example mapper implementation for:
+```java
+public class PokemonMapper implements DbMapper<Pokemon> {
+
+    @Override
+    public Pokemon read(DbRow row) {
+        DbColumn name = row.column("name");
+        DbColumn type = row.column("type");
+        return new Pokemon(name.as(String.class), type.as(String.class));
+    }
+
+    @Override
+    public Map<String, Object> toNamedParameters(Pokemon value) {
+        Map<String, Object> map = new HashMap<>(1);
+        map.put("name", value.getName());
+        map.put("type", value.getType());
+        return map;
+    }
+
+    @Override
+    public List<Object> toIndexedParameters(Pokemon value) {
+        List<Object> list = new ArrayList<>(2);
+        list.add(value.getName());
+        list.add(value.getType());
+        return list;
+    }
+    
+}
+```
+
+## Open questions
+The support for writing rows to WebServer currently transforms the row to JsonObject and then writes it to 
```

Comment:
Are you suggesting changes to the WebServer or simply future enhancements to `DbRowResult` to support these use cases? I read the latter.

Reply:
I meant the first.
Currently web server only supports `Publisher<DataChunk>` - there is no way for us to tell the response that we are sending a publisher of types.
The problem is that the writer needs to understand the type (e.g. `Pokemon.class`) and also understand how it should be send over the network.
For JSON, this would require sending the initial Array definition, then for each `onNext` create a JSON object separated by a comma from the previous one, then in `onComplete` write the end of Array.
I could not wrap my head around this issue and find a nice solution - help please!